### PR TITLE
Refactor for context-awareness

### LINKED ISSUE #1774: Make nodegit context-aware for compatibility with Electron 9 and beyond
[Electron 9 changed](https://www.electronjs.org/blog/electron-9-0#breaking-changes) the default value of `app.allowRendererProcessReuse` to true, which breaks nodegit. Setting it to false manually fixes the issue, but this will only be possible [until Electron 11](https://github.com/electron/electron/blob/9bc5e9823821fc6fef4918f17d591408646b65e5/docs/breaking-changes.md#default-changed-loading-non-context-aware-native-modules-in-the-renderer-process-is-disabled-by-default).

Are there any plans to fix this? Seems like #1298 might be related. Also see https://github.com/electron/electron/issues/18397 for more information.

## System information
  - node version: v12.16.3
  - npm or yarn version: 6.14.4
  - OS/version/architecture: win32 10.0.18363 x64
  - Applicable nodegit version: 0.26.5

```
node -v
npm -v # (or yarn -v)
node -e "console.log(process.platform)"
node -e "console.log(require('os').release())"
node -e "console.log(console.log(process.arch))"
```

### PR DESCRIPTION
**Pull request by ianhattendorf**
TLDR:
- `worker_thread` support!
- Electron support when `app.allowRendererProcessReuse=true`

Fixes #1774

### CODE REVIEW

File: generate/templates/manual/src/thread_pool.cc [hunk 1]
```diff
@@ -672,12 +674,10 @@ namespace nodegit {
       jsThreadCallbackQueue.pop();
     }
 
-    // NOTE We are deliberately leaking this pointer because `async` cleanup in
-    // node has not completely landed yet. Trying to cleanup this pointer
-    // is probably not worth the fight as it's very little memory lost per context
-    // When all LTS versions of node and Electron support async cleanup, we should
-    // be heading back to cleanup this
-    uv_close((uv_handle_t *)jsThreadCallbackAsync, nullptr);
+    uv_close(reinterpret_cast<uv_handle_t *>(jsThreadCallbackAsync), [](uv_handle_t *handle) {
```
Comment (implausible) [inline generate/templates/manual/src/thread_pool.cc:677]: We're leaking the handle still

File: generate/templates/manual/src/thread_pool.cc [hunk 2]
```diff
@@ -672,12 +674,10 @@ namespace nodegit {
...
+      auto threadPoolImpl = static_cast<ThreadPoolImpl *>(handle->data);
+      delete threadPoolImpl->currentContext;
```
Comment (implausible) [inline generate/templates/manual/src/thread_pool.cc:679]: I'm not really a fan of the threadpool deleting the context like this. Could we add a completion callback to the shutdown routine instead?

**Comment by implausible** [comment]
hoooweee...
